src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -406,26 +407,67 @@ public function next_tag( $query = null ) {
 		$breadcrumbs  = $query['breadcrumbs'];
 		$match_offset = isset( $query['match_offset'] ) ? (int) $query['match_offset'] : 1;
 
-		$crumb  = end( $breadcrumbs );
-		$target = strtoupper( $crumb );
 		while ( $match_offset > 0 && $this->step() ) {
-			if ( $target !== $this->get_tag() ) {
-				continue;
+			if ( $this->matches_breadcrumbs( $breadcrumbs ) && 0 === --$match_offset ) {
+				return true;
 			}
+		}
 
-			// Look up the stack to see if the breadcrumbs match.
-			foreach ( $this->state->stack_of_open_elements->walk_up() as $node ) {
-				if ( strtoupper( $crumb ) !== $node->node_name ) {
-					break;
-				}
+		return false;
+	}
 
-				$crumb = prev( $breadcrumbs );
-				if ( false === $crumb && 0 === --$match_offset && ! $this->is_tag_closer() ) {
-					return true;
-				}
+	/**
+	 * Indicates if the currently-matched tag matches the given breadcrumbs.
+	 *
+	 * A "*" represents a single tag wildcard, where any tag matches, but not no tags.
+	 *
+	 * At some point this function _may_ support a `**` syntax for matching any number
+	 * of unspecified tags in the breadcrumb stack. This has been intentionally left
+	 * out, however, to keep this function simple and to avoid introducing backtracking,
+	 * which could open up surprising performance breakdowns.
+	 *
+	 * Example:
+	 *
+	 *     $processor = WP_HTML_Processor::createFragment( '<div><span><figure><img></figure></span></div>' );
+	 *     $processor->next_tag( 'img' );
+	 *     true  === $processor->matches_breadcrumbs( array( 'figure', 'img' ) );
+	 *     true  === $processor->matches_breadcrumbs( array( 'span', 'figure', 'img' ) );
+	 *     false === $processor->matches_breadcrumbs( array( 'span', 'img' ) );
+	 *     true  === $processor->matches_breadcrumbs( array( 'span', '*', 'img' ) );
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string[] $breadcrumbs DOM sub-path at which element is found, e.g. `array( 'FIGURE', 'IMG' )`.
+	 *                              May also contain the wildcard `*` which matches a single element, e.g. `array( 'SECTION', '*' )`.
+	 * @return bool Whether the currently-matched tag is found at the given nested structure.
+	 */
+	public function matches_breadcrumbs( $breadcrumbs ) {
```

Comment:
Maybe this is out of scope for this PR, but could the `$breadcrumbs` format be expanded beyond just a list of tag names to instead be a list of queries (or the like)? That is, it would be very useful to check not just that the tag name matches but more so that the element has a given ID or a class name. Otherwise, the issue is that tag name alone may not be specific enough to be useful.

Reply:
For example, let's say I have the HTML for the entire page and I need to add `fetchpriority=high` to the featured image. In Chrome DevTools, the featured image has the selector `#page > div.single-featured-image-header > img`. In tag name breadcrumbs, this is `['div', 'div', 'img']` which is very unspecific. It would be great to be able to pass in something like:

```php
array(
	array("id" => "page"),
	array(
		"tag" => "div",
		"class" => "single-featured-image-header",
	),
	array("tag" => "img"),
}
```

Reply:
it's definitely out of scope but it's something I've been thinking about a lot recently. it extends to all sorts of challenges in querying and I've been exploring the use of bookmarks to accomplish these needs. one thing we probably don't want to do here is jump around in the document _while jumping around_, meaning either we track everything about every token (memory bloat) or find a way to only trap what we care about when.

the HTML Processor is using `$current_token` to hold not only a tag name, but also a full reference to the location in the document for where something is. this is actually required to properly run the parsing, as certain operations imply closing tags of not only a given name, but a specific tag.

the idea would be something like this, where we still leave this up to userspace so that the performance of the processor remains more predictable, and additional processing/memory can be exposed explicitly in consuming code.

```
while ( $processor->next_tag() ) {
	if (
		'DIV' === $processor->get_tag() &&
		$processor->has_class( 'single-featured-image-header' )
	) {
		$processor->set_bookmark( 'wrapper' );
		continue;
	}

	if (
		'IMG' === $processor->get_tag() &&
		$processor->is_bookmark_open( 'wrapper' )
	) {
		$processor->set_attribute( 'fetchpriority', 'high' );
	}
```

and I believe that we could create a variety of relationship functions like this: `is_child_of( $bookmark_name )`, `is_descendant_of( $bookmark_name )`, `is_sibling_of( $bookmark_name )`.

internally there's a callback for when popping an element off of the stack of open elements, and I have been thinking of exposing that to userspace. I _think_ it should be safe, but I have questions about how best to represent that. this is what led me to the `is_` functions with bookmarks.